packages/block-library/src/paragraph/edit.js

```diff
@@ -64,20 +65,54 @@ function ParagraphBlock( {
 		style: { direction },
 	} );
 
+	const updateContent = useCallback(
+		( newContent ) => setAttributes( { content: newContent } ),
+		[ setAttributes ]
+	);
+
+	const splitParagraph = useCallback(
+		( value, isOriginal ) => {
+			let newAttributes;
+
+			if ( isOriginal || value ) {
+				newAttributes = {
+					...attributes,
+					content: value,
+				};
+			}
+
+			const block = createBlock( name, newAttributes );
+
+			if ( isOriginal ) {
+				block.clientId = clientId;
+			}
+
+			return block;
+		},
+		[ attributes, createBlock ]
+	);
+
+	const updateAlign = useCallback( ( newAlign ) =>
+		setAttributes( { align: newAlign }, [ setAttributes ] )
+	);
+
+	const setDirection = useCallback(
+		( newDirection ) => setAttributes( { direction: newDirection } ),
+		[]
+	);
+
+	const setDropCap = useCallback(
+		() => setAttributes( { dropCap: ! dropCap } ),
+		[ setAttributes ]
```

Comment:
We should add `dropCap` to the dep array too. (I should really pick #24914 up again when I have time 😅 )

Reply:
I have the [same question as above](https://github.com/WordPress/gutenberg/pull/32469/files#r647599862) here; is `setAttributes` changing at some point, making it necessary to add it to the dependency array? As far as I'm aware, only parameters are necessary here.

Reply:
In theory `setAttributes` should be always the same reference set in HOC here:

https://github.com/WordPress/gutenberg/blob/a939c4fbbc639d023ed4a4c747f9f143c29fdef8/packages/block-editor/src/components/block-list/block.js#L231-L242

Reply:
Yep, "in theory". Although it makes no differences when we put it there though. If we enable the eslint plugin, then it will be able to check if we need to add it or not. Yes, there will be false positives, but IMO this is that kind of problems that we should just forward it to automation tools and stop worrying about it anymore.

Reply:
I'm not in favor of declaring dependencies if we're not sure of the effect; it may confuse future modifications. [In this article from the react docs](https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies) they mention that functions _within the scope of the component_ should be included in dependency arrays, but it later also mentions the following:

> **If for some reason you can’t move a function inside an effect, there are a few more options:**
> - You can try moving that function outside of your component. In that case, the function is guaranteed to not reference any props or state, and also doesn’t need to be in the list of dependencies.

I think that `setAttributes` falls under this suggestion, as it is defined outside of the component and it doesn't reference the component's state in its definition.

Something else to note in the [`useCallback`](https://reactjs.org/docs/hooks-reference.html#usecallback) docs is

> [...] Conceptually, though, that’s what they represent: every value referenced inside the callback should also appear in the dependencies array.

I _believe_ that `value` means state within the component, rather than function definitions.

As @gziolo points out, the definition is not expected to change in run-time; `block.js` will never pass down a different definition for `setAttributes` to blocks.

Reply:
I'm really enjoying the conversation here ✨ Just wanted to leave a note that since I dropped `useBlockProps` changes in 9066a9b, I also removed paragraph changes too. (Basically paragraph will always re-render with the style object being updated each time). I would like to revisit `useBlockProps` optimizations, but it sounds like it might be better for a follow up PR.

Reply:
Here is a good summary by React core team: https://github.com/facebook/react/issues/14920#issuecomment-471070149

> First of all, if it's constant and hoisted to top level scope then the linter won't complain. But that doesn't help with things coming from props or context.
>
> If it truly _is_ constant then specifying it in deps doesn't hurt. Such as the case where a `setState` function inside a custom Hook gets returned to your component, and then you call it from an effect. The lint rule isn't smart enough to understand indirection like this. But on the other hand, anyone can wrap that callback later before returning, and possibly reference another prop or state inside it. Then it won’t be constant! And if you fail to handle those changes, you’ll have nasty stale prop/state bugs. So specifying it is a better default.

IMO, it's not about right or wrong, as specifying it or not doesn't change how the code works. It's more of an issue of providing good enough defaults and consensus. The React team officially recommend to use the rule, then for the community sake, we should just follow that, especially when it doesn't break our code in any way. We also get the benefit of not needing to have these discussion in code reviews ever again. In some sense, it's much like the `prettier` evolution to me. Just let the automation tools do their jobs, and stop worrying about them.

It's not really related to this PR anymore though 😅, so feel free to resolve this conversation! I'll bring this up again in #24914 and probably in other channels as well. Good talk as always ❤️!

Reply:
Sure let's resolve this for now, I can try and remember to link back to this conversation if related changes pop up again.

Reply:
@kevin940726 Thanks for that link, it's been insightful and clarifying. I find it really confusing that docs are up to interpretation — which a quick Google search proves, as there's a few articles that discuss this very issue, with varying results — but there's an accepted answer somewhere else.

I agree that it's not about right or wrong, because even those opposites are sort of up to interpretation. In that sense, I suppose it makes sense to stick to what the `eslint` rule enforces, and leave the discussion over agreement with that rule for later, if it ever becomes relevant :)